### ISSUE TITLE
onAttributionChanged listener calls with several processes

### ISSUE THREAD

**Issue by vsochinsky** [adjust/android_sdk#99]
Can the listener be called multiple times if application is running in the several processes? We send event to mixpanel there and need the event to be sent strictly once. Thanks.

**Comment by wellle** [comment]
Yes the `onAttributionChanged` listener will be called exactly once every time the attribution changes. For most devices that will happen exactly once when we have the final attribution data ready after install.

The function will only be called a second time if the attribution changed again at a later point. That would be a reattribution.

See also: https://github.com/adjust/android_sdk/blob/master/doc/mixpanel.md

**Comment by vsochinsky** [comment]
I often have this in logs:
![screen shot 2015-04-18 at 20 10 35](https://cloud.githubusercontent.com/assets/2745764/7216409/4a97f41e-e607-11e4-8999-3ee1c168ee57.png)

In Mixpanel we also see the Marketing Attribution event (tracked in attribution callback) called more than once, sometimes even 3 or more times per user with minor time difference (mostly less than second). What can it be?

**Comment by wellle** [comment]
@vsochinsky: These log lines only mean that the SDK received attribution information multiple times. It does not mean that the listener got called twice. That would only happen if the attribution responses would be different which shouldn't usually be the case.

**Comment by vsochinsky** [comment]
@wellle Ok fine with logs, but we still see lots (actually most) of users in Mixpanel having duplication of the event: 
![screen shot 2015-04-20 at 16 59 45](https://cloud.githubusercontent.com/assets/2745764/7231589/cfa3782c-e77e-11e4-8910-c1961961ff26.png)

This is the ratio of total event calls to number of unique users who called the event:

![screen shot 2015-04-20 at 17 02 37](https://cloud.githubusercontent.com/assets/2745764/7231985/a88b8c22-e781-11e4-88bc-b382f3d20e19.png)

I would be happy to hear any idea about what can be the cause. Everything is implemented just like in the docs and on iOS everything is ok.

**Comment by wellle** [comment]
Ok that is interesting. We'll look into it.

**Comment by vsochinsky** [comment]
@wellle do you have any updates on the issue?

**Comment by uerceg** [comment]
Hi @vsochinsky 

I have few questions for you:

1. Which android_sdk version are you using?

2. What do you mean by "application is running in the several processes" - that your application consists of multiple processes or that somehow your application instance is running in several processes? If you mean this second option, can you explain me exactly how that happens?

3. I see in your log output that you are receiving information that attribution is found in two different processes.
```
29549-29600/com.easyten.android:audio_downloads_service
29473-29618/com.easyten.android
```
Can you explain me why?

Also, if you maybe want to share some code with us, so that I can better understand how does your app work, feel free to write us an email on support@adjust.com.

**Comment by vsochinsky** [comment]
Hi @uerceg !

1. We use the latest version. 4.0.4 when posting logs, 4.0.5 now, and duplication problem still present.

2. I meant that two Application instances are created - one for the app itself, one for background download/processing service which needs to run separately.

3. Adjust sdk is initialized in the Application class. Our app runs in two separate processes, therefore creating two Application instances. Adjust instances somehow not manage to inform each other that attribution was found or something and log this twice.

The Adjust integration code is just the same as in docs. The only thing worth mentioning is that in manifest file services are declared to be running in the separate process (android:process=":audio_downloads_service", in our case).

**Comment by uerceg** [comment]
Hi @vsochinsky,

I see. At this moment, our SDK does not support that you are trying to run different application instances in Android processes with different names. In this implementation of yours, your main process and :audio_downloads_service process are two different processes which don't share same memory space and thus Adjust instance exists in both processes. That's the reason why is your listener for attribution change triggered twice.

At this moment, there's no quick fix from our side which can completely solve this situation, but this is something we are going to work on.

For now, I can only ask you if it's crucial for your application that audio_downloads_service runs in process with different name? If not, if you just remove this property from your manifest file or try to run your service in background thread, SDK should track all events as expected.

**Comment by vsochinsky** [comment]
Hi @uerceg!
Thanks for the info. Yes we can wait for the fix, and yes it's ok to use Adjust only in process of my choice.

**Comment by wellle** [comment]
For reference: It will be implemented in #116.

**Comment by wellle** [comment]
@vsochinsky: Merged #116 and released [v4.0.6](https://github.com/adjust/android_sdk/releases/tag/v4.0.6).

**Comment by vsochinsky** [comment]
@wellle, got it. Will test and if still having issues create new issue.
